### PR TITLE
Don't touch files if they don't need changing

### LINKED ISSUE #88: `numtide/treefmt` compatibility
User report:

> I set up treefmt for my editor and nixfmt fails formatting without any obvious sign.
> Maybe the spec is incomplete, but I think it had to do with some sort of temp-file mutation. iirc.
> Maybe treefmt and nixfmt are competing on the file-handling domain in incompatible ways?

### PR DESCRIPTION
**Pull request by piegamesde**
Fixes #88

This has not been tested yet, CC @SebTM please take a look

### CODE REVIEW

**Comment by SebTM** [comment]
Thanks very much for looking into this as fast as you did 🚀 
Works well, tested it as following:

Modified mtime of flake.nix via `touch -mt 0911171533 flake.nix` and ran new build of nixfmt:
```
-rw-r--r-- 1 sebtm users  3330 Nov 17  2009 flake.nix
dist-newstyle/build/x86_64-linux/ghc-9.0.2/nixfmt-0.5.0/x/nixfmt/build/nixfmt/nixfmt flake.nix
-rw-r--r-- 1 sebtm users  3330 Nov 17  2009 flake.nix
```

running afterwards my old install of nixfmt results in:
```
-rw-r--r-- 1 sebtm users  3330 Nov 17  2009 flake.nix
nixfmt flake.nix
-rw-r--r-- 1 sebtm users  3330 Sep 26 02:15 flake.nix
```

which demonstrates the difference again :) so LGTM

**Review by infinisil — CHANGES_REQUESTED** [review]

File: main/Main.hs [hunk 1]
```diff
@@ -84,8 +85,8 @@ collectAllNixFiles paths = concat <$> mapM collectNixFiles paths
 formatTarget :: Formatter -> Target -> IO Result
 formatTarget format Target{tDoRead, tPath, tDoWrite} = do
     contents <- tDoRead
-    let formatted = format tPath contents
-    mapM tDoWrite formatted
+    let formatResult = format tPath contents
+    mapM (\formatted -> tDoWrite (formatted /= contents) formatted) formatResult
```
[the remark this answers is not in the document]
Comment (infinisil) [reply main/Main.hs:89]: Seems like this might not work after discussing with @piegamesde in the call :)

[the remark this answers is not in the document]
Comment (SebTM) [reply main/Main.hs:89]: Just let me know if I can test again 🥳

**Review by infinisil — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,7 @@
 
 * Fix escaping of interpolations after dollar signs.
 * Fix nixfmt trying to allocate temp files that aren't used.
+* Fix treefmt compatibility
```
Comment (infinisil) [inline CHANGELOG.md:14]:
```suggestion
* Don't write if files didn't change, fixing treefmt compatibility
```

**Review by Lucus16 — APPROVED** [review]

**Comment by infinisil** [comment]
I'd merge, but can't because CI doesn't seem to run

**Comment by piegamesde** [comment]
CI doesn't seem to be running, I have no idea what's happening there. I could probably manually merge and push to master, but that would be undesirable. The CI is broken in #132 too btw

**Comment by Lucus16** [comment]
We'll need to get #134 merged for CI to work. If you don't want to wait for it, just run CI locally and then merge to master manually.

**Pull request merged**